### PR TITLE
feat: Implement user defined table functions

### PR DESCRIPTION
**Pull request by purplefox**
### Description

Implements #3508

This is the 3rd PR in a stack. This is stacked on #3683 so don't review the previous PRs again!

This PR implements user defined table functions. I..e. it allows for the user to provide their own table function implementations in a similar way to how UDFs are implemented using annotations.

Table functions support much of the same functionality as UDFs. They can have multiple parameters which can be of any of the KSQL types. Return values are a List<T> where T is any KSQL type.

The pattern used here is very much the same as that used for UDFs. A lot of files have been touched but that is mainly due to renames.

Schema provider is not supported in this PR and will be the subject of a further PR.

### Testing done
Added some tests to test UDTFs with different types of parameters and return types are loaded.

Added some more QTT tests
### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    final List<Schema> args = ImmutableList.of(STRUCT_SCHEMA);
+
+    // When:
+    final KsqlTableFunction function = FUNC_REG.getTableFunction("test_udtf", args);
+
+    // Then:
+    assertThat(function.getReturnType(), equalTo(STRUCT_SCHEMA));
+  }
+
+  private static final Schema STRUCT_SCHEMA =
```
Comment (agavra) [inline ksql-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java]: style: can we put constants up top?

File: ksql-udf/src/main/java/io/confluent/ksql/function/udtf/UdtfDescription.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2018 Confluent Inc.
```
Comment (agavra) [inline ksql-udf/src/main/java/io/confluent/ksql/function/udtf/UdtfDescription.java]: nit: dates

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java [hunk 3]
```diff
@@ -167,134 +154,134 @@ private void loadUdfs(final ClassLoader loader, final Optional<Path> path) {
...
-      final UdfDescription annotation,
-      final Method method,
-      final String path
-  ) {
-    LOGGER.info("Adding UDF name='{}' from class={}", annotation.name(), theClass);
-    final UdfInvoker invoker = createUdfInvoker(method);
-    addFunction(theClass, annotation, method, invoker, path);
+    functionRegistry.ensureFunctionFactory(factory);
+
+    Arrays.stream(theClass.getMethods())
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: I tend to agree. I'm not usually a fan of using streams for the sake of it. This is a copy and paste (old code) but I will change it.

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java [hunk 4]
```diff
@@ -167,134 +154,134 @@ private void loadUdfs(final ClassLoader loader, final Optional<Path> path) {
...
-            classLevelAnnotation.version(),
-            path,
-            false)));
 
-    final List<Schema> parameters = IntStream.range(0, method.getParameterCount()).mapToObj(idx -> {
+    final List<Schema> parameters = createParameters(method, functionName);
+
+    final Schema javaReturnSchema = getReturnType(method, udfAnnotation.schema());
+
+    return KsqlFunction.create(
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: +1, would be clearer

File: ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    return explode(input);
+  }
+
+  @Udtf
+  public List<String> explodeString(final List<String> input) {
+    return explode(input);
+  }
+
+  @Udtf
+  public List<BigDecimal> explodeBigDecimal(final List<BigDecimal> input) {
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java]: SchemaProvider is coming in a later PR. But I think the reason it works, is that the check that disallows BigDecimal code without a schemaProvider exists in the schema Provider code, so we don't have that check in the code right now.

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: I think I would rather make the change in a later PR, the reason being that in a later PR I split out the UdfLoader into multiple loader classes, and if I change here I'll just have to reapply the change anyway in the later PR.

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: Will apply this as part of the review of https://github.com/confluentinc/ksql/pull/3689 as it gets factored out anyway (as above)

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java [hunk 6]
```diff
@@ -320,44 +308,82 @@ private void addFunction(final Class theClass,
...
+        = Arrays.stream(theClass.getMethods())
+        .filter(method -> method.getAnnotation(UdafFactory.class) != null)
+        .filter(method -> {
+          if (!Modifier.isStatic(method.getModifiers())) {
+            LOGGER.warn(
+                "Trying to create a UDAF from a non-static factory method. Udaf factory"
+                    + " methods must be static. class={}, method={}, name={}",
+                method.getDeclaringClass(),
+                method.getName(),
+                udafAnnotation.name()
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: as above

File: ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java [hunk 7]
```diff
@@ -320,44 +308,82 @@ private void addFunction(final Class theClass,
...
+                udafAnnotation.name(),
+                method.getName(),
+                method.getDeclaringClass(),
+                path,
+                e
+            );
+          }
+          return Optional.<UdafFactoryInvoker>empty();
+        }).filter(Optional::isPresent)
+        .map(Optional::get)
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/UdfLoader.java]: as above

File: ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java]: I will remove the tests for BigDecimal from this PR

[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java]: done

File: ksql-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java (hunk 1 above)
Comment (purplefox) [reply ksql-engine/src/test/java/io/confluent/ksql/function/UdtfLoaderTest.java]: done

File: ksql-udf/src/main/java/io/confluent/ksql/function/udtf/UdtfDescription.java (hunk 2 above)
Comment (purplefox) [reply ksql-udf/src/main/java/io/confluent/ksql/function/udtf/UdtfDescription.java]: done

File: ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udtf/array/Explode.java]: talked offline - this will fail hard in the future PR as well

**Pull request merged**